### ISSUE TITLE
feat: normalize errors with h3's `createError`

### ISSUE THREAD

**Issue by Sandros94** [Intevel/nuxt-directus#243]
I've yet to test this, but instead of doing a bunch of `console.error`s I could serialize any error coming from Directus.

Will update once I have time to work on this

**Comment by zguig52** [comment]
Hello, still me here :). I updated to latest version of the module to try evolutions and so on ("^0.0.10"). Default user profile request works like a charm, thanks a lot for this!

I have questions/concerns about errors management so it might be the right place here.

Concerning errors management for the login method: I see that the useDirectusAuth().login method catch all errors, print them in the console and return a null value:
   ```
 try {
      const authResponse = await client
        .request(sdkLogin(identifier, password, defu(options, { mode: defaultMode })))

      if (updateStates !== false) {
        if (updateTokens !== false) {
          await setTokens(authResponse)
        }
        if (readMe !== false) {
          await readMyself(readMe)
        }
      }

      return authResponse
    } catch (error: any) {
      if (error && error.message) {
        console.error("Couldn't login user.", error.message)
      } else {
        console.error(error)
      }
      return null
    }
  }
```

I don't know if/how you have planned to change this but I think errors causes/details have to be thrown in one way or another, because we need to know for example that the password is wrong, based on returned payload or that the server is down or whatever kind of error has happened to warn the user about it and/or handle the issue.

This behaviour seems to be also found in other methods like useDirectusItems().createItem where I can also only 'blindly' throw a message and not analyse and handle it properly based on returned response if any.

Why not just throwing the raw error instead of replying with null for the moment in all final catch methods?

Is the h3's createError evolution planned to take care of all this?

**Comment by Sandros94** [comment]
> Is the h3's createError evolution planned to take care of all this?

Exactly, but since I wanted to investigate a little bit more how errors are generated in the SDK I decided to postpone it to the next minor release

**Comment by zguig52** [comment]
I just found a very very annoying use case due to the fact of not throwing errors:
==> useDirectusItems().deleteItem

I have exactly the same behavior, that it work or not, as the return value is undefined either if it works or not...

I am not able to handle, even blindly the deletion error.

**Comment by zguig52** [comment]
And the same might happen for "dark hole" items, where you only have creation permissions and no read permission, as Directus will reply with a 204 without any body.

**Comment by Sandros94** [comment]
> [...] as the return value is undefined either if it works or not...
>
> I am not able to handle, even blindly the deletion error.

there are a number of functions from the SDK, in particular the `delete*` and `logout` that do return nothing, requiring a doublecheck via other means if that item was successfully deleted.

On the other hand, while working on `useAsyncData` support I've noticed that the number of `try`-`catch` blocks that I had to use before being able to use `$fetch` was still there. And since I'm using `$fetch` directly from Nuxt I don't have to worry about adding nor managing errors caused by fetch itself (nor retries or similar).

I've already did commit some changes on this topic, but I'll need to finish some other few before releasing a new `nuxt-directus-next` update. I'll update this issue once I do.

### Side effect

The SDK has a few errors that are not currently properly handled, blocking UX.
Since I see an open issue on this ([directus/#21474](https://github.com/directus/directus/issues/21474)) I might not just map each error one by one and wait for any update on that side (or at least until we are ready for a stable `6.0` release).

**Comment by Sandros94** [comment]
@zguig52 I've updated `nuxt-directus-next` to `0.0.11`, [here](https://github.com/Intevel/nuxt-directus/discussions/215#discussioncomment-8574579) you can find some notes about the changes.

Now we should be able to handle all the errors caused during a fetch, with the only handful of generic errors generated by the SDK for things like missing `ID`s or trying to operate on core collections.

Based on the ETA and direction of the Directus issue I'll either close this for now or leave it open, but feel free to write if you encounter any error that could be caused by me (in that case remember to provide a reproduction).

**Comment by zguig52** [comment]
That's fantastic, thanks a lot for this!
Here are my feedbacks and tests based on 0.11 release. I've been mainly playing with directus permissions (403 + data with errors details from directus) and server stop (proxy in front, empty 503 reply). 

Included in a try catch:
- useDirectusAuth().login OK
- useDirectusAuth().logout OK
- useDirectusFiles().deleteFiles OK
- useDirectusFiles().uploadFiles OK
- useDirectusItems().deleteItem OK
- useDirectusItems().readItem OK

Usage:
```
try{
METHOD CALL HERE
}catch(error: any){
  console.error("functionName", error)
  displayMessage("error", error.data?.errors[0]?.message ?? error) // whatever you want to process the error
}
```

With async syntax (const { data, error}):
- useDirectusItems().readAsyncItems OK

Usage:
```
    const { data, error } = await useDirectusItems<Collections>().readAsyncItems(collection, {
      query: {
        fields: fields
      }
    })
    if(!error.value && data.value){
      return data.value;
    }else{
      console.error("getExistingRegistration", error.value)
      displayMessage("error", error.value.data?.errors[0]?.message ?? error) // whatever you want to process the error
    }
```

One remark, when using the readItems, query parameter signature are a bit different from the readAsyncItems:
readItems(collection, {fields: ['*']}) (has changed with recent version if I remember well - was like the async signature today before)
readAsyncItems(collection, {query: {fields: ['*']}}) (did not put the ref stuff to make the difference more visible)
==> It might be better to align both with the same query syntax for easier use, don't you think?

BTW, thanks for the example on how to type the different parameters of functions in your release notes, it helped me a lot to remove many //@ts-ignore :)!

**Comment by zguig52** [comment]
> If yes then this is intentional for a number of reasons:

Thanks for the clarifications!

**Comment by spyesx** [comment]
Hi, thx for the work. Any head up on this issue?

I'm trying to handle the following error but I can't capture it.

```js
{
    "errors": [
        {
            "message": "Value for field \"email\" in collection \"directus_users\" has to be unique.",
            "extensions": {
                "collection": "directus_users",
                "field": "email",
                "code": "RECORD_NOT_UNIQUE"
            }
        }
    ]
}
```

**Comment by sandros94** [comment]
> Hi, thx for the work. Any head up on this issue?

@spyesx this issue mainly focuses on the next major version (v6), but could you provide a code snippet of your use-case? In theory you should be able to catch that already.

**Comment by spyesx** [comment]
I understand. I found this morning only some docs about the v6. What is the official branch for it?

My use case is a very simple registration:

```js
  const { register } = useDirectusAuth();

  try {
    const newUser = await createUser({ email: data.email, password: data.password });

    if(newUser?.errors) {
    // Can't catch things here. Right?
    }
  } catch (e) {
    console.log(e)
    // Can't catch things here.
  }
```

**Comment by sandros94** [comment]
> I understand. I found this morning only some docs about the v6. What is the official branch for it?

[dev](https://github.com/Intevel/nuxt-directus/tree/dev) is the official one, although I stared working on [nuxt-first](https://github.com/Intevel/nuxt-directus/tree/nuxt-first) for a no-sdk approach, but is undocumented and unreleased (as we are still considering the ideal approach of this module).

Feel free to leave a comment in https://github.com/Intevel/nuxt-directus/discussions/271 is you have any ideas

> My use case is a very simple registration:
> 
>```js
>const { register } = useDirectusAuth();
> 
>try {
>  const newUser = await createUser({ email: data.email, password: data.password });
>
>  if(newUser?.errors) {
>    // Can't catch things here. Right?
>  }
>} catch (e) {
>  console.log(e)
>  // Can't catch things here.
>}
>```

I see both `register` and `createUser`, which one is the actual one?
Keep in mind that if you are importing the official SDK then it has its own [`createUser` that needs to be wrapped in a `client.request`](https://directus.io/docs/api/users#create-an-user)

**Comment by spyesx** [comment]
> [dev](https://github.com/Intevel/nuxt-directus/tree/dev) is the official one, although I stared working on [nuxt-first](https://github.com/Intevel/nuxt-directus/tree/nuxt-first) for a no-sdk approach, but is undocumented and unreleased (as we are still considering the ideal approach of this module).
> 
> Feel free to leave a comment in https://github.com/Intevel/nuxt-directus/discussions/271 is you have any ideas

Thx :)

Actually, I am using the composable from this  nuxt-directus and [`register` is an alias for `createUser`](https://github.com/Intevel/nuxt-directus/blob/706fa0482b14944d49874e967d3d1ae27daff1cb/src/runtime/composables/useDirectusAuth.ts#L120). I took the option to go for `createUser` because I don't need the alias. Maybe catching the error is possible but I can't get my head around it.
